### PR TITLE
feat(source): column pruning for parquet file source

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
For Parquet encode, we can get it's own schema before reading whole file, thus we can do column pruning to read only the columns that the user needs by setting `ProjectionMask`.

The same optimization can also be done on tvf(`file_scan(paruqet_file)`), but since tvf and batch source will be unified in the backend, this optimization is only implemented on the file source side.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by wcy-fdu** [comment]
Kindly call for review❤️cc @chenzl25 @hzxa21

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 1]
```diff
@@ -57,14 +62,15 @@ impl<Src: OpendalSource> SplitReader for OpendalReader<Src> {
         splits: Vec<OpendalFsSplit<Src>>,
         parser_config: ParserConfig,
         source_ctx: SourceContextRef,
-        _columns: Option<Vec<Column>>,
```
Comment (chenzl25) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:60]: Previously, the columns variable was unused, does it mean it is None?

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 2]
```diff
@@ -215,3 +229,60 @@ impl<Src: OpendalSource> OpendalReader<Src> {
...
+                                .ok()?;
+                            if &arrow_field == converted_arrow_schema.field(pos) {
+                                Some(pos)
+                            } else {
+                                None
+                            }
+                        })
+                })
+                .collect();
+            valid_column_indices.sort();
```
Comment (chenzl25) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:283]: It seems the order of the input columns isn't respected, but the order matters. Considering parquet schema is ['a', 'b', 'c'], but the column order we want is [`c`, `b`]. The `valid_column_indices.sort()` here will lead to [`b`, `c`] which is unexpected.

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs (hunk 1 above)
Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:60]: No, it's just not used. Here the columns is Some(`user_defined_columns`, `hidden_columns`).

I just run some test to verify
```
CREATE Table test_s3(
        id int,
        age int,
    ) WITH (
        connector = 's3',
        match_pattern = '*.parquet',
    s3.region_name = 'custom',
    s3.bucket_name = 'hummock001',
    s3.credentials.access = 'hummockadmin',
    s3.credentials.secret = 'hummockadmin',
    s3.endpoint_url = 'http://hummock001.127.0.0.1:9301',
    ) FORMAT PLAIN ENCODE PARQUET;
```
and the columns here is `Some([Column { name: "id", data_type: Int32, is_visible: true }, Column { name: "age", data_type: Int32, is_visible: true }, Column { name: "_row_id", data_type: Serial, is_visible: false }, Column { name: "_rw_s3_v2_file", data_type: Varchar, is_visible: false }, Column { name: "_rw_s3_v2_offset", data_type: Varchar, is_visible: false }])`

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:283]: The `valid_column_indices` is used to construct `ProjectionMask`, and in  `ProjectionMask`, repeated or out of order indices will not impact the final mask. i.e. `[0, 1, 2]` will construct the same mask as `[1, 0, 0, 2]`. 
So no matter what order, there will be no problem with correctness. But I don't mind removing the sort here.

Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs]: Previously I keep `.sort()` just refer to how [iceberg-rust](https://github.com/apache/iceberg-rust/blob/cafc939d6aa303f7424a74fa8467ac0cb306591d/crates/iceberg/src/arrow/reader.rs#L378) does, let me also send a pr to remove this in iceberg-rust😄.

Comment (chenzl25) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs]: I see. `ParquetParser` will reorder the chunk returned from `ParquetRecordBatchStream`.

**Review by chenzl25 — APPROVED** [review]
LGTM

**Pull request merged**